### PR TITLE
Improve experiments / notifications system to more flexible

### PR DESCRIPTION
**Pull request by kpethtel**
**Story card:** [ch3514](https://app.clubhouse.io/simpledotorg/story/3514/can-we-run-this-as-an-experiment)

This now does a few things to make our notification system flexible enough to handle the med refill messaging we want to send out, as well as still allowing support for appointment reminders. It changes Notification to have a polymorphic relationship to a `subject`, which for Appt reminders would maintain the appointment relationship.

It also adds a backfill so that Communication data is copied over to Notification...this will allow us to remove redundant relationships in the future.  

There are also a bunch more renames going on to get things consistent here.

* [x] remove dependency on `Appointment` from `Communication`
* [x] rename `AppointmentReminder` to `Notification`
* [x] make `appointment` optional on `Notification`
* [x] rename any other related models / locals / etc
* [x] green build to verify all the things!
* [ ] verify we can send these notifications to Hari's phone on SBX

### CODE REVIEW

File: app/models/appointment.rb
Comment (rsanheim) [inline app/models/appointment.rb:149]: I don't think this quite matches the previous logic ?

File: db/migrate/20210517215935_remove_appointment_from_communication_and_add_patient.rb
Comment (prabhanshuguptagit) [inline db/migrate/20210517215935_remove_appointment_from_communication_and_add_patient.rb:4]: Before we drop `appointment_id`, should we add a data migration to copy over `patient_id` for existing `communications`?

**Review by prabhanshuguptagit — COMMENTED** [review]
Looking great so far 💯 .

File: app/models/appointment.rb
Comment (kpethtel) [inline app/models/appointment.rb:69]: We might have to make some changes to accommodate this. The reason I added the "where no communications" is to exclude all past appointments. Given that Notification is a new model, this will result in all past appointments receiving new notifications. This could also be fixed by changing
```
def self.overdue_by(number_of_days)
  overdue.where("scheduled_date <= ?", Date.current - number_of_days.days)
end
```
to only search for the scheduled date = date. But that would mean we might never send a reminder in some edge cases (notifications were turned off for a while, sidekiq dropped jobs).

File: app/models/appointment.rb
Comment (kpethtel) [reply app/models/appointment.rb:149]: I think it achieves the same thing. Unrelated, but I think "attempted" is a poorly named method. It's defined as 
```
  def attempted?
    successful? || in_progress?
  end
  ```
  which means failures weren't attempted. So it's really more like "not failed".

File: config/locales/api/am_ET.yml
Comment (kpethtel) [inline config/locales/api/am_ET.yml:17]: I'm the one who did this, but I'm not sure we needed to change these translations. They are actually appointment reminders.

File: db/migrate/20210517215935_remove_appointment_from_communication_and_add_patient.rb
Comment (kpethtel) [reply db/migrate/20210517215935_remove_appointment_from_communication_and_add_patient.rb:4]: We briefly considered that but didn't pursue it. Maybe we should, then we wouldn't have to make patient optional for the notification model.

**Comment by kpethtel** [comment]
We should also consider renaming "reminder_template" to "notification_template" or something like that.

File: db/migrate/20210517215935_remove_appointment_from_communication_and_add_patient.rb
Comment (rsanheim) [reply db/migrate/20210517215935_remove_appointment_from_communication_and_add_patient.rb]: I've now added a backfill.

File: db/schema.rb
Comment (rsanheim) [inline db/schema.rb:144]: I no longer think we need `patient_id` on communication, because notifications has all that data.

File: app/models/appointment.rb
Comment (kpethtel) [inline app/models/appointment.rb:152]: I'd prefer this with a leading `!!` since this is a `?` method.

File: app/services/appointment_notification_service.rb
Comment (kpethtel) [inline app/services/appointment_notification_service.rb:35]: I still think the locale keys are probably the only thing we should leave as is.

File: app/services/backfill_notifications.rb
Comment (kpethtel) [inline app/services/backfill_notifications.rb:1]: Nice. I think this will save us some pain.

File: db/migrate/20210518182324_make_notifications_polymorphic_to_subject.rb
Comment (kpethtel) [inline db/migrate/20210518182324_make_notifications_polymorphic_to_subject.rb:5]: Does this also need a `subject_type`? I suppose this works at least until we add a second subject type.

File: db/schema.rb
Comment (kpethtel) [reply db/schema.rb:144]: I think you're right.

File: lib/tasks/scripts/purge_users_data.rb
Comment (kpethtel) [inline lib/tasks/scripts/purge_users_data.rb:8]: Good catch.

File: spec/services/backfill_notifications_spec.rb
Comment (kpethtel) [inline spec/services/backfill_notifications_spec.rb:3]: I would suggest we delete these sometime in the future, since this script should only be used once and there's no reason to test it forever.

File: db/migrate/20210518182324_make_notifications_polymorphic_to_subject.rb
Comment (rsanheim) [reply db/migrate/20210518182324_make_notifications_polymorphic_to_subject.rb:5]: This adds the subject_id and subject_type automatically.

Comment (kpethtel) [reply db/migrate/20210518182324_make_notifications_polymorphic_to_subject.rb:5]: I see. Missed that in the schema.

File: app/services/backfill_notifications.rb
Comment (kpethtel) [inline app/services/backfill_notifications.rb:5]: I think this needs to be changed since you changed the locales key back to "appointment_reminders".

File: db/data/20210518183334_create_notifications_for_communications.rb
Comment (kpethtel) [inline db/data/20210518183334_create_notifications_for_communications.rb:7]: I don't think it provides any real value, but we could just destroy all notifications as a means of reversing this. But this is probably the better choice.

File: lib/seed/runner.rb
Comment (kpethtel) [inline lib/seed/runner.rb:11]: This won't turn it on in Prod, right? I must not understand the relationship between these defaults and our prod flags.

File: spec/jobs/appointment_notification/worker_spec.rb
Comment (kpethtel) [inline spec/jobs/appointment_notification/worker_spec.rb:14]: This should also probably be changed back to appointment_reminders.

**Review by kpethtel — COMMENTED** [review]
Great work, very thorough!

**Comment by kpethtel** [comment]
Unfortunately, I can't approve this because I opened it, but you have my approval.

File: app/services/backfill_notifications.rb
Comment (rsanheim) [reply app/services/backfill_notifications.rb]: oh right, fixed

File: db/data/20210518183334_create_notifications_for_communications.rb
Comment (rsanheim) [reply db/data/20210518183334_create_notifications_for_communications.rb:7]: This is just the default that the data migrate tool gives us, because making them reversible is usually not worth the trouble

File: lib/seed/runner.rb
Comment (rsanheim) [reply lib/seed/runner.rb:11]: Correct, this is just seed data, so will only impact dev and sandbox (when we rerun seeds there).

File: spec/jobs/appointment_notification/worker_spec.rb
Comment (rsanheim) [reply spec/jobs/appointment_notification/worker_spec.rb]: oh ya fixed

File: spec/services/backfill_notifications_spec.rb
Comment (rsanheim) [reply spec/services/backfill_notifications_spec.rb:3]: 👍

**Review by rsanheim — APPROVED** [review]

**Pull request merged**
